runtime/runtime/src/tests/apply.rs

```diff
@@ -2793,3 +2837,186 @@ fn test_deploy_and_call_local_receipts() {
+    let alice_signer = InMemorySigner::test_signer(&alice_account());
+    let bob_signer = InMemorySigner::test_signer(&bob_account());
+    let alice_invalid_signer = InMemorySigner::from_seed(alice_account(), KeyType::ED25519, "seed");
+
+    // This transaction should be droped due to invalid signer.
+    let alice_invalid_tx = SignedTransaction::send_money(
+        1,
+        alice_account(),
+        alice_account(),
+        &alice_invalid_signer,
+        100,
+        CryptoHash::default(),
+    );
+    let alice_tx1 = SignedTransaction::send_money(
+        1,
+        alice_account(),
+        alice_account(),
+        &alice_signer,
+        200,
+        CryptoHash::default(),
+    );
+    let alice_tx2 = SignedTransaction::send_money(
+        2,
+        alice_account(),
+        bob_account(),
+        &alice_signer,
+        300,
+        CryptoHash::default(),
+    );
+    let bob_tx1 = SignedTransaction::send_money(
+        1,
+        bob_account(),
+        bob_account(),
+        &bob_signer,
+        400,
+        CryptoHash::default(),
+    );
+    let bob_tx2 = SignedTransaction::send_money(
+        2,
+        bob_account(),
+        alice_account(),
+        &bob_signer,
+        500,
+        CryptoHash::default(),
+    );
+    let bob_tx3 = SignedTransaction::send_money(
+        3,
+        bob_account(),
+        bob_account(),
+        &bob_signer,
+        600,
+        CryptoHash::default(),
+    );
+
+    let txs = vec![
+        bob_tx1.clone(),
+        alice_invalid_tx,
+        alice_tx1.clone(),
+        bob_tx2.clone(),
+        alice_tx2.clone(),
+        bob_tx3.clone(),
+    ];
+
+    let (runtime, tries, root, apply_state, _signers, epoch_info_provider) = setup_runtime(
+        vec![alice_account(), bob_account()],
+        to_yocto(1_000_000),
+        to_yocto(500_000),
+        10u64.pow(15),
+    );
+
+    let validity_flags = vec![true; txs.len()];
+    let signed_valid_period_txs =
+        SignedValidPeriodTransactions::new(&txs, validity_flags.as_slice());
+    let apply_result = runtime
+        .apply(
+            tries.get_trie_for_shard(ShardUId::single_shard(), root),
+            &None,
+            &apply_state,
+            &[],
+            signed_valid_period_txs,
+            &epoch_info_provider,
+            Default::default(),
+        )
+        .expect("apply should succeed");
+
+    let expected_order = vec![
+        bob_tx1.get_hash(),
+        alice_tx1.get_hash(),
+        bob_tx2.get_hash(),
+        alice_tx2.get_hash(),
+        bob_tx3.get_hash(),
+    ];
+
+    // Note: The 3 local receipts are generated for valid transactions
+    // where signer_id == receiver_id - tx2, tx4, tx6 (not for tx1 as it is dropped).
+    assert_eq!(
+        apply_result.outcomes.len(),
+        8,
+        "should have processed 5 transactions and 3 local receipts"
+    );
+    let tx_outcomes = apply_result.outcomes.iter().take(5).map(|o| o.id).collect::<Vec<_>>();
```

Comment:
this is interesting.  I think we are also asserting the order of the receipts here.  Is that guaranteed?

Reply:
actually, I ignore the receipts here (`.take(5)` only takes outcomes of transactions)